File: 2022/day11/part1/main.py
import operator
ops = {"+": operator.add, "*": operator.mul}


def processInput(input):
    data = {"total": 0}
    arr = input.split("\n")

    # process Monkey
    val = arr[0].strip().split("Monkey ")
    data["monkey"] = int(val[1].strip().split(":")[0])

    # process Starting Items
    val = arr[1].strip().split("Starting items: ")
    val = val[1].strip().split(", ")
    data["items"] = list(map(int, val))

    # process Operation
    val = arr[2].strip().split("Operation: new = old ")
    val = val[1].strip().split(" ")
    data["operation"] = lambda a: ops[val[0]](
        a, int(val[1])) if val[1] != "old" else ops[val[0]](a, a)

    # process Test
    test = int(arr[3].strip().split("Test: divisible by ")[1])
    true = int(arr[4].strip().split("If true: throw to monkey ")[1])
    false = int(arr[5].strip().split("If false: throw to monkey ")[1])
    data["test"] = lambda a: true if a % test == 0 else false

    return data


with open('./input.txt') as f:
    result = 0
    rounds = 1
    monkeys = []
    bufferedInput = ""
    for line in f:
        if line == "\n":
            value = processInput(bufferedInput)
            monkeys.append(value)
            bufferedInput = ""
        else:
            bufferedInput += line
    value = processInput(bufferedInput)
    monkeys.append(value)
    bufferedInput = ""

    while rounds <= 20:
        for monkey in monkeys:
            while len(monkey["items"]):
                monkey["total"] += 1
                item = monkey["items"].pop(0)
                item = monkey["operation"](item)
                item = item // 3
                goto = monkey["test"](item)
                monkeys[goto]["items"].append(item)
        rounds += 1

    def cond(a): return a["total"]
    monkeys.sort(key=cond, reverse=True)
    result = monkeys[0]["total"] * monkeys[1]["total"]

    print(result)

File: 2022/day11/part1/main.js
const fs = require("fs");

function processInput(input) {
  const ops = {
    "+": (a, b) => a + b,
    "*": (a, b) => a * b,
  };
  let res = [];
  let monkey = {};
  for (let idx = 0; idx < input.length; idx += 7) {
    monkey["total"] = 0;
    monkey["monkey"] = parseInt(input[idx].split("Monkey ")[1].split(":")[0]);
    monkey["items"] = input[idx + 1]
      .split("Starting items: ")[1]
      .split(",")
      .map((val) => parseInt(val.trim()));
    let temp = input[idx + 2].split("Operation: new = ")[1];
    monkey["operation"] = (val) => {
      return eval(temp.replaceAll("old", val));
    };
    const testVal = parseInt(input[idx + 3].split("Test: divisible by ")[1]);
    monkey["testVal"] = testVal;
    const ifTrue = parseInt(
      input[idx + 4].split("If true: throw to monkey ")[1]
    );
    const ifFalse = parseInt(
      input[idx + 5].split("If false: throw to monkey ")[1]
    );

    monkey["test"] = (val) => {
      return val % testVal == 0 ? ifTrue : ifFalse;
    };
    res.push(monkey);
    monkey = {};
  }
  return res;
}

let result = 0;
let rounds = 1;
let monkeys = [];

try {
  const data = fs
    .readFileSync("input.txt", "utf8")
    .split("\n")
    .map((line) => line.trim());
  monkeys = processInput(data);
} catch (err) {
  console.error(err);
}
const lcm = monkeys.reduce((a, b) => a * b.testVal, monkeys[0].testVal);

while (rounds <= 20) {
  monkeys.forEach((monkey) => {
    while (monkey.items.length) {
      monkey.total++;
      let item = monkey.items.shift();
      item = monkey.operation(item);
      item = Math.floor(item / 3);
      let pass = monkey.test(item);
      monkeys[pass].items.push(item);
    }
  });
  rounds++;
}

monkeys.sort((a, b) => b.total - a.total);
result = monkeys[0]["total"] * monkeys[1]["total"];

console.log(result);
